go-controller/pkg/crd/userdefinednetwork/v1/shared.go

```diff
@@ -128,16 +128,44 @@ type Layer2Config struct {
 	// +optional
 	JoinSubnets DualStackCIDRs `json:"joinSubnets,omitempty"`
 
-	// IPAMLifecycle controls IP addresses management lifecycle.
+	// IPAM section contains IPAM-related configuration for the network.
+	// +optional
+	IPAM *IPAMSpec `json:"ipam,omitempty"`
+}
+
+// +kubebuilder:validation:XValidation:rule="!has(self.lifecycle) || self.lifecycle != 'Persistent' || !has(self.mode) || self.mode == 'Enabled'", message="lifecycle Persistent is only supported when ipam.mode is Enabled"
+// +kubebuilder:validation:MinProperties=1
+type IPAMSpec struct {
+	// Mode controls how much of the IP configuration will be managed by OVN.
+	// `Enabled` means OVN-Kubernetes will apply IP configuration to the SDN infrastructure and it will also assign IPs
+	// from the selected subnet to the individual pods.
+	// `Disabled` means OVN-Kubernetes will only assign MAC addresses and provide layer 2 communication, letting users
+	// configure IP addresses for the pods.
+	// `Disabled` is only available for Secondary networks.
+	// By disabling IPAM, any Kubernetes features that rely on selecting pods by IP will no longer function
+	// (such as network policy, services, etc). Additionally, IP port security will also be disabled for interfaces attached to this network.
+	// Defaults to `Enabled`.
+	// +optional
+	Mode IPAMMode `json:"mode,omitempty"`
```

Comment:
pardon my ignorance, but couldn't we set a default using kubebuilder ?

Reply:
you cannot change the default value without breaking the API afaiu.

Reply:
you kinda can if you can figure out a "smooth upgrade" https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#defaulting